File: typescript/src/modules/token.ts
/**
 * API Token管理服务模块
 * 基于 @sttot/axios-api 的标准化API定义
 */

import { apiBase } from '@sttot/axios-api';
import { commonErrorHandler } from '../utils/errorHandler';
import {
  CreateApiTokenRequest,
  CreateApiTokenResponse,
  UpdateApiTokenRequest,
  DeleteApiTokenRequest,
  ListApiTokensRequest,
  ListApiTokensResponse,
  ApiToken,
  EmptyResponse
} from '../types';

export const initTokenService = (apiFactory: ReturnType<typeof apiBase>) => {
  class TokenService {
    // 创建API Token
    public createApiToken = apiFactory<CreateApiTokenRequest, CreateApiTokenResponse>(
      (request) => ({
        method: 'POST',
        url: '/api/v1/tokens',
        data: {
          name: request.name,
          role: request.role,
          description: request.description,
          expire_days: request.expireDays
        }
      }),
      ({ data }) => ({
        tokenInfo: {
          id: data.token_info.id,
          name: data.token_info.name,
          role: data.token_info.role,
          description: data.token_info.description,
          active: data.token_info.active,
          expireAt: data.token_info.expire_at,
          lastUsedAt: data.token_info.last_used_at,
          lastUsedIp: data.token_info.last_used_ip,
          createdAt: data.token_info.created_at,
          updatedAt: data.token_info.updated_at
        },
        tokenValue: data.token_value
      }),
      commonErrorHandler
    );

    // 获取API Token列表
    public listApiTokens = apiFactory<ListApiTokensRequest, ListApiTokensResponse>(
      (request = {}) => ({
        method: 'GET',
        url: '/api/v1/tokens',
        params: this.buildParams({
          page: request.page,
          page_size: request.pageSize,
          search: request.search
        })
      }),
      ({ data }) => ({
        tokens: data.tokens.map((token: any) => ({
          id: token.id,
          name: token.name,
          role: token.role,
          description: token.description,
          active: token.active,
          expireAt: token.expire_at,
          lastUsedAt: token.last_used_at,
          lastUsedIp: token.last_used_ip,
          createdAt: token.created_at,
          updatedAt: token.updated_at
        })),
        total: data.total,
        page: data.page,
        pageSize: data.page_size
      }),
      commonErrorHandler
    );

    // 获取API Token详情
    public getApiToken = apiFactory<{ id: number }, ApiToken>(
      (request) => ({
        method: 'GET',
        url: `/api/v1/tokens/${request.id}`
      }),
      ({ data }) => ({
        id: data.id,
        name: data.name,
        role: data.role,
        description: data.description,
        active: data.active,
        expireAt: data.expire_at,
        lastUsedAt: data.last_used_at,
        lastUsedIp: data.last_used_ip,
        createdAt: data.created_at,
        updatedAt: data.updated_at
      }),
      commonErrorHandler
    );

    // 更新API Token
    public updateApiToken = apiFactory<UpdateApiTokenRequest, ApiToken>(
      (request) => ({
        method: 'PUT',
        url: `/api/v1/tokens/${request.id}`,
        data: {
          name: request.name,
          role: request.role,
          description: request.description,
          active: request.active
        }
      }),
      ({ data }) => ({
        id: data.id,
        name: data.name,
        role: data.role,
        description: data.description,
        active: data.active,
        expireAt: data.expire_at,
        lastUsedAt: data.last_used_at,
        lastUsedIp: data.last_used_ip,
        createdAt: data.created_at,
        updatedAt: data.updated_at
      }),
      commonErrorHandler
    );

    // 删除API Token
    public deleteApiToken = apiFactory<DeleteApiTokenRequest, EmptyResponse>(
      (request) => ({
        method: 'DELETE',
        url: `/api/v1/tokens/${request.id}`
      }),
      ({ data }) => data as EmptyResponse,
      commonErrorHandler
    );

    // 辅助方法：构建查询参数，过滤掉undefined值
    public buildParams(params: Record<string, any>): Record<string, any> {
      const result: Record<string, any> = {};
      for (const [key, value] of Object.entries(params)) {
        if (value !== undefined && value !== null) {
          result[key] = value;
        }
      }
      return result;
    }
  }

  return new TokenService();
};

File: golang/modules/token.go
package modules

import (
	"context"
	"strconv"

	"github.com/NewNanCity/NewNanManager-SDK/clients/golang/utils"
	"github.com/go-resty/resty/v2"
)

// TokenService Token管理服务
type TokenService struct {
	client *resty.Client
}

// NewTokenService 创建Token服务实例
func NewTokenService(client *resty.Client) *TokenService {
	return &TokenService{client: client}
}

// ListApiTokens 获取API Token列表
func (s *TokenService) ListApiTokens() (*ListApiTokensData, error) {
	return s.ListApiTokensWithContext(context.Background())
}

// ListApiTokensWithContext executes ListApiTokens with cancellation scoped to this request.
func (s *TokenService) ListApiTokensWithContext(ctx context.Context) (*ListApiTokensData, error) {
	return s.listApiTokens(ctx, nil)
}

// ListApiTokensPage 获取指定页的API Token列表，Page从1开始，PageSize范围1-100。
func (s *TokenService) ListApiTokensPage(request ListApiTokensRequest) (*ListApiTokensData, error) {
	return s.ListApiTokensPageWithContext(context.Background(), request)
}

// ListApiTokensPageWithContext 在调用方上下文中获取指定页的API Token列表。
func (s *TokenService) ListApiTokensPageWithContext(ctx context.Context, request ListApiTokensRequest) (*ListApiTokensData, error) {
	return s.listApiTokens(ctx, &request)
}

func (s *TokenService) listApiTokens(ctx context.Context, request *ListApiTokensRequest) (*ListApiTokensData, error) {
	req := s.client.R().SetContext(ctx)
	if request != nil {
		req.SetQueryParam("page", strconv.Itoa(int(request.Page)))
		req.SetQueryParam("page_size", strconv.Itoa(int(request.PageSize)))
	}
	resp, err := req.Get("/api/v1/tokens")

	var result ListApiTokensData
	err = utils.HandleResponse(resp, err, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// CreateApiToken 创建API Token
func (s *TokenService) CreateApiToken(request CreateApiTokenRequest) (*CreateApiTokenData, error) {
	return s.CreateApiTokenWithContext(context.Background(), request)
}

// CreateApiTokenWithContext executes CreateApiToken with cancellation scoped to this request.
func (s *TokenService) CreateApiTokenWithContext(ctx context.Context, request CreateApiTokenRequest) (*CreateApiTokenData, error) {
	resp, err := s.client.R().SetContext(ctx).
		SetBody(request).
		Post("/api/v1/tokens")

	var result CreateApiTokenData
	err = utils.HandleResponse(resp, err, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// GetApiToken 获取API Token详情
func (s *TokenService) GetApiToken(id int32) (*ApiToken, error) {
	return s.GetApiTokenWithContext(context.Background(), id)
}

// GetApiTokenWithContext executes GetApiToken with cancellation scoped to this request.
func (s *TokenService) GetApiTokenWithContext(ctx context.Context, id int32) (*ApiToken, error) {
	resp, err := s.client.R().SetContext(ctx).
		Get("/api/v1/tokens/" + strconv.Itoa(int(id)))

	var result ApiToken
	err = utils.HandleResponse(resp, err, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// UpdateApiToken 更新API Token
func (s *TokenService) UpdateApiToken(id int32, request UpdateApiTokenRequest) (*ApiToken, error) {
	return s.UpdateApiTokenWithContext(context.Background(), id, request)
}

// UpdateApiTokenWithContext executes UpdateApiToken with cancellation scoped to this request.
func (s *TokenService) UpdateApiTokenWithContext(ctx context.Context, id int32, request UpdateApiTokenRequest) (*ApiToken, error) {
	resp, err := s.client.R().SetContext(ctx).
		SetBody(request).
		Put("/api/v1/tokens/" + strconv.Itoa(int(id)))

	var result ApiToken
	err = utils.HandleResponse(resp, err, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// DeleteApiToken 删除API Token
func (s *TokenService) DeleteApiToken(id int32) error {
	return s.DeleteApiTokenWithContext(context.Background(), id)
}

// DeleteApiTokenWithContext executes DeleteApiToken with cancellation scoped to this request.
func (s *TokenService) DeleteApiTokenWithContext(ctx context.Context, id int32) error {
	resp, err := s.client.R().SetContext(ctx).
		Delete("/api/v1/tokens/" + strconv.Itoa(int(id)))

	return utils.HandleResponse(resp, err, nil)
}
